src/v0/sources/shopify/util.js

```diff
@@ -98,52 +99,96 @@ const getCartToken = (message) => {
+ * @param {*} message 
+ */
+const getRudderIdFromNoteAtrributes = (noteAttributes, field) => {
+  const rudderIdObj = noteAttributes.find(obj => obj.name === field);
+  if (isDefinedAndNotNull(rudderIdObj)) {
+    return rudderIdObj.value;
   }
-  return generateUUID();
-};
+  return null;
+}
 /**
- * This function sets the anonymousId based on cart_token or id from the properties of message.
- * If it's null then we set userId as "shopify-admin".
+ * This function gets the anonymousId based on cart_token from redis
  * @param {*} message
  * @returns
  */
-const getAnonymousIdFromDb = async (message, metricMetadata) => {
-  const cartToken = getCartToken(message);
-  if (isDefinedAndNotNull(cartToken)) {
-    let anonymousId;
-    stats.increment('shopify_redis_calls', {
+const getAnonymousIdFromDb = async (cartToken, message, metricMetadata) => {
+  let anonymousId;
+  stats.increment('shopify_redis_calls', {
+    type: 'get',
+    ...metricMetadata,
+  });
+  try {
+    anonymousId = await RedisDB.getVal(`${cartToken}`, 'anonymousId');
+  } catch (e) {
+    stats.increment('shopify_redis_failures', {
       type: 'get',
       ...metricMetadata,
     });
-    try {
-      anonymousId = await RedisDB.getVal(`${cartToken}`, 'anonymousId');
-    } catch (e) {
-      stats.increment('shopify_redis_failures', {
-        type: 'get',
-        ...metricMetadata,
-      });
-    }
-    if (isDefinedAndNotNull(anonymousId)) {
-      return anonymousId;
+  }
+  if (isDefinedAndNotNull(anonymousId)) {
+    return anonymousId;
+  }
+  stats.increment('shopify_redis_no_val', {
+    ...metricMetadata,
+    event: message.event,
+  });
+  // If there is no mapping or there is some error in connection we return null
+  return null;
+};
+
+/**
+ * This function retrieves anonymousId in folowing steps:
+ * 1. Checks for `rudderAnonymousId` in `note_atrributes`
+ * 2. if redis is enabled checks in redis
+ * 3. This means we don't have `anonymousId` and hence events CAN NOT be stitched and we check for cartToken
+ *    a. if cartToken is available we return its hash value
+ *    b. else we check if the event is an SHOPIFY_ADMIN_ONLY_EVENT 
+ *       -> if true we return `null`;
+ *       -> else we don't have any identifer (very edge case) we return `random anonymousId`
+ * @param {*} message 
+ * @param {*} metricMetadata 
+ * @returns 
+ */
+const getAnonymousId = async (message, metricMetadata) => {
+  let anonymousId;
+  const noteAttributes = message.properties?.note_attributes;
+  // Giving Priority to note_attributes to fetch rudderAnonymousId over Redis ue to better functionality
+  if (isDefinedAndNotNull(noteAttributes)) {
+    anonymousId = getRudderIdFromNoteAtrributes(noteAttributes, "rudderAnonymousId");
+  }
+  // falling back to cartToken mapping or its hash in case no rudderAnonymousId is found
+  if (!isDefinedAndNotNull(anonymousId)) {
+    const cartToken = getCartToken(message);
+    if (isDefinedAndNotNull(cartToken)) {
```

Comment:
if (!isDefinedAndNotNull(cartToken)) {
   if (SHOPIFY_ADMIN_ONLY_EVENTS.includes(message.event)) {
        return null;
      }
      return generateUUID();
}

this will save another nested if

Reply:
Done